### ISSUE TITLE
Using rules in params?

### ISSUE THREAD

**Issue by mikebaldry** [hanami/hanami#1265]
Great job on Hanami and congrats on this big release!

I'm trying to make a custom rule for a `Hanami::Action::Params` subclass but it doesn't appear to be execute them. 

I couldn't find an example using rules, and tried various syntax until this one didn't fail on boot, but placing a `pry` inside doesn't stop and the tests fail.

```ruby
class Register < Hanami::Action::Params
  params do
    required(:email).filled(:str?)
    required(:password).filled(:str?, min_size?: 8, max_size?: 50)

    rules do
      rule(:email) do
        unless value.include?("@")
          key.failure("must be a valid email address")
        end
      end
    end
  end
end
```

What is the correct way to apply slightly more complex validation at this level?

I also was looking at custom predicates so I could do `required(:email).filled(:str?, :email?)` but I can't work out how to add new ones as I wasn't able to find any documentation on that, I assume there is some way to register them or have them be picked up with auto inject?

Keep up the good work

**Comment by solnic** [comment]
@mikebaldry you can define a contract class and just pass it as an argument to `params` method.

**Comment by jodosha** [comment]
@mikebaldry For more info about contracts: https://dry-rb.org/gems/dry-validation/1.8/schemas/

**Comment by solnic** [comment]
I should clarify that by `params` method I meant `Hanami::Action.params` method 🙂

**Comment by mikebaldry** [comment]
> @mikebaldry you can define a contract class and just pass it as an argument to `params` method.

So I've defined a contract, and I want to use it in an `Hanami::Action`, so i use:

```ruby
params MyContract

def handle(request, response)
  # now request.params = an instance of MyContract, which I need to call call on and pass in the actual params
  # but the params are no longer accessible, because they've been overwritten by the instance of MyContract
  # I can access the raw params via request.env, but that doesn't seem right?
end
```

I can of course stop using `params` and just do `half 422 unless MyContract.new.call(request.params).success?` but that bypasses the whole action thing.

**Comment by solnic** [comment]
@katafrakt sorry for bothering you but I'm pretty sure you figured out how to use a full-blown dry-validation contract for params validation, would you mind sharing your solution? 🙏🏻

**Comment by katafrakt** [comment]
I do, as a matter of fact. This is my code:

```ruby
class MyApp::Action
  def self.contract(&block)
    klass = Class.new(Hanami::Action::Params)
    contract = Dry::Validation::Contract.build { instance_eval(&block) }
    klass.instance_variable_set(:@_validator, contract)
    @params_class = klass # this could as well be `params klass`
  end
end
```

and then in an action itself:

```ruby
class Account::Actions::Registration::Create < Account::Action
  contract do
    schema do
      required(:email).filled(:str?, format?: URI::MailTo::EMAIL_REGEXP)
      required(:password).filled(:str?, min_size?: 8)
      required(:password_confirmation).filled(:str?, min_size?: 8)
    end

    rule(:password_confirmation, :password) do
      key.failure("passwords do not match") if values[:password] != values[:password_confirmation]
    end
  end

  def handle(req, res)
    if !req.params.valid?
      # [...]
    end
  end
end
```

The suggesion to define a contract class and pass it using a `params` class method did not work for me as well. It seems to expect a subclass of `Hanami::Action::Params`.

**Comment by wuarmin** [comment]
@jodosha @solnic I think we should fix this. What do you think? The workaround of @katafrakt works, but it is difficult to solve this way, especially for new users. What do you think?

**Comment by solnic** [comment]
@wuarmin we're going to address this in 2.x

**Comment by marcin-mantke-ext-solaris** [comment]
@solnic Is it anywhere on the roadmap? Any ETA we could expect this?

**Comment by solnic** [comment]
@marcin-mantke-ext-solaris it will be done eventually, can't give you an ETA, sorry

**Comment by timriley** [comment]
We have an open PR for this in https://github.com/hanami/controller/pull/434.

I'd like to make sure this is included in the 2.2 release.

**Comment by timriley** [comment]
This will now be implemented by https://github.com/hanami/controller/pull/453, which will merge soon, and will definitely be included in the upcoming 2.2 release.
